mteb/abstasks/AbsTaskClusteringFast.py

```diff
@@ -140,3 +144,75 @@ def _evaluate_monolingual(
         )
 
         return {"v_measures": v_measures, "v_measure": float(np.mean(v_measures))}
+
+
+def clustering_downsample(
+    dataset: DatasetDict, seed: int, max_samples_in_cluster: int = 2048
+) -> DatasetDict:
+    """In cases where it is not possible to convert the dataset to a fast version, we can downsample the dataset to speed up the evaluation.
+
+    This might be necessary when the clusters in the dataset is not sampled from the same distribution.
+    """
+    rng_state = random.Random(seed)
+
+    ds = {}
+    for split in dataset:
+        _docs = []
+        _labels = []
+
+        n_clusters = len(dataset[split])
+
+        for i in range(n_clusters):
+            labels = dataset[split]["labels"][i]
+            sentences = dataset[split]["sentences"][i]
+
+            n_sample = min(max_samples_in_cluster, len(sentences))
+
+            # sample n_sample from each cluster
+            idxs = rng_state.sample(range(len(sentences)), n_sample)
+            _docs.append([sentences[idx] for idx in idxs])
+            _labels.append([labels[idx] for idx in idxs])
+
+        ds[split] = Dataset.from_dict({"sentences": _docs, "labels": _labels})
+    return DatasetDict(ds)
+
+
+def convert_to_fast(
+    dataset: DatasetDict, seed: int, max_size: int = 100_000
+) -> DatasetDict:
+    """Converts a clustering dataset to a fast version. This concats the cluster into two columns, sentences and labels.
+    It additionally downsamples the dataset to max_size.
+    """
```

Comment:
The idea is to use this for conversion and then the dataset can potentially be re-uploaded to save on the downloading time.